### DISCUSSION TITLE
Geospatial operatos for mongodb

### DISCUSSION THREAD

**Discussion by World3Developer** [mikro-orm/mikro-orm#3872] [Q&A]
We need spatial operators in mongo queries.  

$box, $polygon, ... must be incorporated if the driver is mongodb.  (Or may be a parameter in options, or something easy to implement)

The actual problem is that if the "key" is not supported, $in is added if the value is an array.  This can be avoided if $box, $polygon, ... were in the QueryHelper.SUPPORTED_OPERATORS.

QueryHelper.SUPPORTED_OPERATORS = ['>', '<', '<=', '>=', '!', '!=', ':in', ':nin', ':gt', ':gte', ':lt', ':lte', ':ne', ':not', '$box', '$polygon'];

How can I get this behaviour without modifying the actual code of the library?

Thanks in advance.

**Comment by B4nan** [comment]
You should be able to use any native query via `em.getCollection()` which gives you the native mongo collection object - and map the raw results back to entities via `em.map()`.

In v6 we could probably find more native ways to do this, I will think about that. 

> How can I get this behaviour without modifying the actual code of the library?

Given the `QueryHelper.SUPPORTED_OPERATORS` is a static array, you could just append to it, it should work just fine.

```ts
import { QueryHelper } from '@mikro-orm/core';

QueryHelper.SUPPORTED_OPERATORS.push('$box');
```

**Comment by World3Developer** [comment]
Sorry if I misunderstand your explanation.  I tried this, but got same result.  The find query was modified with the operator $in.

```
const coll = this.entityManager.getCollection(CellEntity);

    const res = coll.find({
      coordinates: {
        $geoWithin: {
          $geometry: {
            type: 'Polygon',
            coordinates: [
              [0, 0],
              [0, 100],
              [100, 0],
              [100, 100],
            ],
          },
        },
      },
    });

    console.log(res);

    const entities = res.map((row) => this.entityManager.map(CellEntity, row));
```

**Reply by B4nan** [reply]
Well, I doubt it. As said, you work with the native mongo collection object - there is no ORM involved in your code. What you describe happens if you do `em.find()`, not with `em.getCollection().find()`.

**Reply by World3Developer** [reply]
This was a nice trick:

```
import { QueryHelper } from '@mikro-orm/core';
QueryHelper.SUPPORTED_OPERATORS.push('$box');
```

You can close this issue, but may be comment in documentation.  I lost a lot of time trying that.  

Thanks a lot.

**Reply by World3Developer** [reply]
> Well, I doubt it. As said, you work with the native mongo collection object - there is no ORM involved in your code. What you describe happens if you do `em.find()`, not with `em.getCollection().find()`.

I tried to do that:

```
    const cellsInBB = this.entityManager.getCollection(CellEntity).find({
      coordinates: {
        $geoWithin: {
          $box: bbox,
        },
      },
    } as any);
```

But it returns a FindCursor<WithId<Document>> and I don't know what must do with it.

**Comment by World3Developer** [comment]
You should be able to use any native query via em.getCollection() which gives you the native mongo collection object - and map the raw results back to entities via em.map().

@B4nan Where can I find an example to do that?

I have complex aggregate functions that have the same problem with the operators.

```
    const results = this.entityManager
      .getCollection(GroupEntity)
      .aggregate(query);

    console.log(results);
```

Type of results is AggregationCursor<Document>, but I don't know how to operate with it.

**Reply by B4nan** [reply]
As said, you operate with the _native mongo collection_, so read the docs for the node mongo driver on how to use it. 

https://www.mongodb.com/docs/drivers/node/current/
